urle381b6cba7_IkarosKappler_ikrs_http_tgz-pJ8-ikrs_json_rpc_JSONRPCHandlerJ8/src/ikrs/json/parser/JSONBuilderExample.java

```diff
@@ -0,0 +1,278 @@
+     * @throws NullPointerException If reader is null.
+     **/
+    public JSONBuilderExample( Reader reader,
+			       boolean caseSensitive
+			       ) 
+	throws NullPointerException {
+	
+	super( reader, caseSensitive );
+	
+	this.objectStack      = new Stack<Object>();
+	this.typeStack        = new Stack<Integer>();
+	this.memberNameStack  = new Stack<String>();
+	
+	this.objectStack.push( new Object() ); // dummy
+	this.typeStack.push( new Integer(-1) );
+	this.memberNameStack.push( null );
+	this.currentType = -1;
+    }    
+    
+    private Object popFromStack() {
+
+	// Remove from stack
+	this.lastPoppedObject = this.objectStack.pop();
+	this.typeStack.pop();	
+	this.currentMemberName = this.memberNameStack.pop();
+
+	this.currentType = this.typeStack.peek().intValue();
+
+	
+	this.currentArray = null;
+	this.currentObject = null;
+	if( this.currentType == JSONTYPE_ARRAY ) 
+	    this.currentArray = (List<Object>)this.objectStack.peek();
+	else if( this.currentType == JSONTYPE_OBJECT ) 
+	    this.currentObject = (Map<String,Object>)this.objectStack.peek();    
+	else 
+	    ; // NOOP
+
+	this.currentMemberValue = this.lastPoppedObject;
+
+	return this.lastPoppedObject;
+    }
+
+    //--- BEGIN --------- Override parser event methods ---------- //
+    @Override
+    protected void fireJSONBegin() {	
+    }
+
+    @Override
+    protected void fireJSONEnd() {	
+    }
+
+    @Override
+    protected void fireArrayBegin() {	
+	this.currentType = JSONTYPE_ARRAY;
+	this.currentArray = new LinkedList<Object>();
+	this.objectStack.push( this.currentArray );
+	this.typeStack.push( new Integer(JSONTYPE_ARRAY) );
+	this.memberNameStack.push( this.currentMemberName );
+    }
+
+    @Override
+    protected void fireArrayElementEnd() {
+	this.currentArray.add( this.currentMemberValue );
+    }
+    
+    @Override
+    protected void fireArrayEnd() {
+	this.popFromStack();	
+    }
+
+    @Override
+    protected void fireObjectBegin() {
+	this.currentType = JSONTYPE_OBJECT;
+	this.currentObject = new TreeMap<String,Object>();
+	this.objectStack.push( this.currentObject );
+	this.typeStack.push( new Integer(JSONTYPE_OBJECT) );
+	this.memberNameStack.push( this.currentMemberName );
+    }
+
+    @Override
+    protected void fireObjectEnd() {
+	this.popFromStack();	
+    }
+
+    @Override
+    protected void fireMemberBegin() {
+	// NOOP
+    }
+
+    @Override
+    protected void fireMemberNameRead( String token ) {
+	this.currentMemberName = token;
+    }
+
+    @Override
+    protected void fireMemberEnd() {
+	this.currentObject.put( currentMemberName, 
+				this.currentMemberValue
+				);
+	
+    }
+
+    @Override
+    protected void fireNumberRead( String number ) {
+	try {
+	    try {
+		// Try to parse Integer
+		this.currentMemberValue = new Integer(number);
+	    } catch( NumberFormatException e ) {
+		this.currentMemberValue = new Double(number);
+	    }
+	} catch( NumberFormatException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)